### PR TITLE
Feat/mp/restendpoints

### PR DESCRIPTION
**Pull request by markkopaas**

### CODE REVIEW

File: src/data-source.handlers.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import {dataSourcesPath} from "./data-source.router";
+
+interface ResourceHandlerMap {
+  [key: string]: RequestHandler;
+}
+
+export const dataSourceHandlers: ResourceHandlerMap = {
+  create: (req, res, next) => {
+    const data = req.body;
+    data.id = uuidv4(); //todo domain logic should be in domain class, see if it is possible to do it in mongoose
```
Comment (Chlenix) [inline src/data-source.handlers.ts]: denne kan settes inn i schema object

File: src/data-source.model.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  dataSourceType?: string;
+  url?: string;
+  publisherId?: string;
+  description?: string;
+}
+
+export interface DataSourceDocument extends Document, DataSource {}
+
+const dataSourceSchemaDefinition = {
+  id: String,
```
Comment (Chlenix) [inline src/data-source.model.ts]: auto-generere id

```
import uuidv4 from 'uuid/v4';
const dataSourceSchemaDefinition = {
    id: { type: String, default: uuidv4 }
    // ...
};
```

File: src/db/db.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import mongoose from 'mongoose';
+import config from 'config';
+
+mongoose.Promise = global.Promise;
+
+const logRuntimeError = (error: mongoose.Error): void => {
+  console.error(error);
+};
```
Comment (azaxarov) [inline src/db/db.ts]: Redundant function

File: src/db/db.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+  const options = {
+    useNewUrlParser: true,
+    useUnifiedTopology: true
+  };
+
+  mongoose
+    .connect(url, options)
+    .then(() => {
+      mongoose.connection.on('error', logRuntimeError);
```
Comment (azaxarov) [inline src/db/db.ts]: mongoose.connection.on('error', **console.error**);

File: src/db/db.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    })
+    .catch(abortAndExit);
```
Comment (azaxarov) [inline src/db/db.ts]: Could also use async await with try/catch

File: src/db/db.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import mongoose from 'mongoose';
+import config from 'config';
+
+mongoose.Promise = global.Promise;
```
Comment (azaxarov) [inline src/db/db.ts]: mongoose does not support native promises?

File: src/data-source.model.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  dataSourceType: String,
+  url: String,
+  publisherId: String,
+  description: String
+};
+
+const dataSourceFromDocument = (data: DataSourceDocument): DataSource =>
+  pick(data, Object.getOwnPropertyNames(dataSourceSchemaDefinition));
```
Comment (azaxarov) [inline src/data-source.model.ts]: Object.keys?

File: src/data-source.model.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+const DataSourceSchema = new mongoose.Schema(dataSourceSchemaDefinition, {
```
Comment (azaxarov) [inline src/data-source.model.ts]: Perhaps import Schema from mongoose?

File: src/data-source.model.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  timestamps: true,
+  strict: true,
+  toObject: {
+    transform: (_doc, ret): any => dataSourceFromDocument(ret)
+  }
+});
+
+export const DataSourceModel = mongoose.model<DataSourceDocument>(
```
Comment (azaxarov) [inline src/data-source.model.ts]: Perhaps import model from mongoose?

File: src/data-source.handlers.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  NotFoundHttpError
+} from './lib/normalize-http-error';
+import uuidv4 from 'uuid/v4';
+import {dataSourcesPath} from "./data-source.router";
+
+interface ResourceHandlerMap {
+  [key: string]: RequestHandler;
+}
+
+export const dataSourceHandlers: ResourceHandlerMap = {
```
Comment (azaxarov) [inline src/data-source.handlers.ts]: Perhaps a class implementation would be cleaner?

File: src/app.ts [hunk 11]
```diff
@@ -2,15 +2,21 @@ import express from 'express';
...
+
 app.use(commonErrorHandler);
 
 app.listen(8000, err => {
-  console.log('server running');
   if (err) {
-    return console.log(err);
+    return console.error(err);
   }
 });
```
Comment (azaxarov) [inline src/app.ts]: Consider implementing an async function (e.g. run()), that will first instantiate mongoose and then call app.listen using async await syntax.

File: docker-compose.yml [hunk 12]
```diff
@@ -0,0 +1,12 @@
+version: "3.5"
+
+services:
+  # TODO: node
+
+  mongodb:
+    image: mongo:latest
+    environment:
+      - MONGO_INITDB_ROOT_USERNAME=$FDKHARVESTADM_MONGO_USERNAME
+      - MONGO_INITDB_ROOT_PASSWORD=$FDKHARVESTADM_MONGO_PASSWORD
```
[the remark this answers is not in the document]
Comment (markkopaas) [reply docker-compose.yml]: skal fikse

File: config/default.json [hunk 13]
```diff
@@ -0,0 +1,7 @@
+{
+  "mongodb": {
+    "host": "localhost",
+    "port": 27017,
+    "name": "admin"
+  }
+}
```
[the remark this answers is not in the document]
Comment (Chlenix) [reply config/default.json]: nope, it's the "default" profile name of config.js - https://www.npmjs.com/package/config

File: src/data-source.handlers.ts (hunk 10 above)
Comment (Chlenix) [reply src/data-source.handlers.ts]: It used to be a class with inheritance. I second that it should still be a class (but with static handlers this time and no inheritance)

File: src/db/db.ts (hunk 6 above)
Comment (Chlenix) [reply src/db/db.ts]: Looks like it's legacy code from Mongoose 4 (We're using 5). I'll remove it

**Pull request merged**